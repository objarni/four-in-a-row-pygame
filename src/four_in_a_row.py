import pygame

from src.messages import LeftMouseDownAt, LeftMouseUpAt, MouseMovedTo, Tick
from src.printers import print_model
from src.states import StartScreenState
from src.constants import WIDTH, HEIGHT, FPS

from src.update import update
from src.view import view


def main():
    global api
    pygame.init()
    pygame.mixer.init()
    screen = pygame.display.set_mode((WIDTH, HEIGHT))
    pygame.display.set_caption("Four in a row")
    resource_folder = 'res'
    drawing_api = DrawingAPI(screen, resource_folder)
    audio_api = AudioAPI(resource_folder)
    mainloop(drawing_api, audio_api)
    pygame.quit()


def mainloop(drawing_api, audio_api):
    model = StartScreenState()
    view(model, drawing_api)
    pygame.display.update()
    clock = pygame.time.Clock()
    while True:
        old_model_repr = print_model(model)

        # A tick happens every time around the loop!
        model = update(model, Tick(pygame.time.get_ticks()), audio_api)

        # Translate low level events to domain events
        while ev := pygame.event.poll():
            msg = None
            if ev.type == pygame.MOUSEBUTTONDOWN:
                if ev.button == 1:
                    msg = LeftMouseDownAt(ev.pos)
            if ev.type == pygame.MOUSEBUTTONUP:
                if ev.button == 1:
                    msg = LeftMouseUpAt(ev.pos)
            if ev.type == pygame.MOUSEMOTION:
                msg = MouseMovedTo(ev.pos)
            if ev.type == pygame.KEYDOWN:
                if ev.key == pygame.K_q:
                    return
            elif ev.type == pygame.QUIT:
                return

            if msg:
                model = update(model, msg, audio_api)

        # Display current model, if any change found
        if old_model_repr != print_model(model):
            view(model, drawing_api)

        pygame.display.update()
        clock.tick(FPS)


class DrawingAPI:
    def __init__(self, screen, resource_path):
        self.screen = screen
        self.image_dict = {}
        self.resource_path = resource_path
        self.font = pygame.font.Font(f'{self.resource_path}/font.ttf', 30)

    def draw_rectangle(self, center, size, color):
        r = pygame.Rect(0, 0, *size)
        r.center = center
        pygame.draw.rect(self.screen, color, r)

    def draw_disc(self, center, size, color):
        pygame.draw.circle(self.screen, color, center, size, size)

    def draw_text(self, center, text, color):
        text_surface = self.font.render(text, True, color)
        text_rect = text_surface.get_rect()
        text_rect.center = center
        self.screen.blit(text_surface, text_rect)

    def draw_image(self, center, name, dimension):
        if (name, dimension) not in self.image_dict:
            self.image_dict[(name, dimension)] = self.load_and_scale(name, dimension)
        image = self.image_dict[(name, dimension)]
        pos = center[0] - WIDTH // 2, center[1] - HEIGHT // 2
        self.screen.blit(image, pos)

    def load_and_scale(self, name, dimension):
        p = f'{self.resource_path}/{name}.png'
        image = pygame.image.load(p)
        return pygame.transform.scale(image, dimension)


class AudioAPI:
    def __init__(self, resource_path):
        self.resource_path = resource_path
        self.sound = {}

    def play_music(self, name):
        p = f'{self.resource_path}/{name}.ogg'
        pygame.mixer.music.load(p)
        pygame.mixer.music.play(-1)

    def stop_music(self):
        pygame.mixer.music.stop()

    def play_sound(self, name):
        if name not in self.sound:
            p = f'{self.resource_path}/{name}.wav'
            self.sound[name] = pygame.mixer.Sound(p)
        self.sound[name].play()


if __name__ == '__main__':
    main()

# TODO: all states as named tuples instead of classes and initial_GameState function
# TODO: refactor to many states instead of 'mixed in' states
# TODO: little man easter egg
# TODO: remove old_model_repr logic (animations on every screen so was premature optimization after all, but fun exp!)
# TODO: ALSA blocks mixer init. Can it be solved without resorting to 'catch exception / null audio api'?
# TODO: idea to reduce missing updating projectors with new state:
#         investigate using nametuple string representation in projectors


