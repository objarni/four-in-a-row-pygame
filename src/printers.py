from src.constants import EMPTY, RED, YELLOW, ROWS, COLUMNS
from src.states import StartScreenState, GameOverState, GameState
from src.update import print_color


def print_model(model):
    state_string = model.__class__.__name__ + '\n'
    if isinstance(model, StartScreenState):
        state_string += f'{model.time=}\n'
        state_string += f'{model.music_playing=}\n'
    if isinstance(model, GameOverState):
        state_string += print_board(model.board)
        state_string += f'\n{print_color(model.winner).title()} won.\n'
    if isinstance(model, GameState):
        state_string += f'It is {print_color(model.whos_turn_is_it)}s turn.\n'
        state_string += f'{model.time=}\n'
        state_string += f'The mouse is at {model.mouse_pos}.\n'
        state_string += f'{model.mouse_down_time=}\n'
        state_string += print_board(model.board)
    return state_string


def print_board(board):
    board_string = ''
    symbols = {
        EMPTY: 'O',
        RED: 'R',
        YELLOW: 'Y'
    }
    board_string += "0 1 2 3 4 5 6\n"
    board_string += "-------------\n"
    for y in range(ROWS):
        board_string += ' '.join(symbols[board[(x, y)]] for x in range(COLUMNS)) + '\n'
    return board_string