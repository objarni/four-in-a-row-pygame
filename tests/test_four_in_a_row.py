
import pygame
import pytest
from approvaltests import verify

import src.constants
import src.four_in_a_row
import src.states
import src.update
from src.messages import LeftMouseDownAt, ColumnWasClicked, MouseMovedTo
from src.states import GameState, GameOverState, StartScreenState
from tests.fakes import simulate_main_event_loop
from tests.printers import print_rgb, print_scenario, SimulationLog


@pytest.fixture
def log():
    logger = SimulationLog()
    src.update.log = logger
    return logger


@pytest.fixture
def game_underway():
    return GameState()


def setup_function():
    pygame.init()


def test_print_rgb():
    assert print_rgb(0, 0, 0) == '.'
    assert print_rgb(200, 0, 0) == 'R'
    assert print_rgb(0, 200, 0) == 'G'
    assert print_rgb(0, 0, 200) == 'B'
    assert print_rgb(200, 200, 0) == 'Y'
    assert print_rgb(200, 200, 200) == 'W'


def test_first_placed_brick_is_red(log):
    game = GameState()
    player_moves = [ColumnWasClicked(0)]
    result = simulate_main_event_loop(game, player_moves, log)
    verify(print_scenario(player_moves, result, log))


def test_placing_4_bricks_in_first_column(log):
    game = GameState()
    player_moves = [ColumnWasClicked(0)] * 4
    result = simulate_main_event_loop(game, player_moves, log)
    verify(print_scenario(player_moves, result, log))


def test_placing_too_many_bricks_in_first_column(log):
    game = GameState()
    player_moves = [ColumnWasClicked(0)] * 8
    result = simulate_main_event_loop(game, player_moves, log)
    verify(print_scenario(player_moves, result, log))


def test_letting_red_win(log):
    game = GameState()
    player_moves = [ColumnWasClicked(0), ColumnWasClicked(1)] * 3 + [ColumnWasClicked(0)]
    result = simulate_main_event_loop(game, player_moves, log)
    verify(print_scenario(player_moves, result, log))


def test_letting_red_win_horisontally(log):
    game = GameState()
    player_moves = [ColumnWasClicked(c) for c in [0, 5, 1, 5, 2, 5, 3]]
    result = simulate_main_event_loop(game, player_moves, log)
    verify(print_scenario(player_moves, result, log))


def test_letting_yellow_win_horisontally(log):
    game = GameState()
    player_moves = [ColumnWasClicked(c) for c in [0, 1, 0, 2, 0, 3, 1, 4]]
    result = simulate_main_event_loop(game, player_moves, log)
    verify(print_scenario(player_moves, result, log))


def test_slash_red_win(log):
    game = GameState()
    player_moves = [ColumnWasClicked(c) for c in [0, 1, 1, 2, 2, 3, 2, 3, 3, 5, 3]]
    result = simulate_main_event_loop(game, player_moves, log)
    verify(print_scenario(player_moves, result, log))


def test_backslash_yellow_win(log):
    game = GameState()
    player_moves = [ColumnWasClicked(c) for c in [0, 6, 5, 5, 4, 4, 3, 4, 5, 3, 0, 3, 0, 3]]
    result = simulate_main_event_loop(game, player_moves, log)
    verify(print_scenario(player_moves, result, log))


def test_startscreen(log):
    game = StartScreenState()
    ticks = [src.messages.Tick(ms) for ms in range(3)]
    result = simulate_main_event_loop(game, ticks, log)
    verify(print_scenario(ticks, result, log))


def test_clicking_in_game_over_state(log):
    game = GameOverState(winner=src.constants.RED, board=src.states.empty_board())
    mouse_events = [LeftMouseDownAt((1, 1))]
    result = simulate_main_event_loop(game, mouse_events, log)
    verify(print_scenario(mouse_events, result, log))


def test_startscreen_to_game_transition(log):
    game = StartScreenState()
    mouse_events = [LeftMouseDownAt((1, 1))]
    result = simulate_main_event_loop(game, mouse_events, log)
    verify(print_scenario(mouse_events, result, log))


def test_mouse_movement_over_game_screen(log):
    game = GameState()
    mouse_events = [MouseMovedTo((x, x)) for x in range(0, 500, 10)]
    result = simulate_main_event_loop(game, mouse_events, log)
    verify(print_scenario(mouse_events, result, log))
