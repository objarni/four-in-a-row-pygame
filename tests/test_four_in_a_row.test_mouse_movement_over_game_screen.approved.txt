SCENARIO:
mouse_movement_over_game_screen
events:
MouseMovedTo(pos=(0, 0))
MouseMovedTo(pos=(10, 10))
MouseMovedTo(pos=(20, 20))
MouseMovedTo(pos=(30, 30))
MouseMovedTo(pos=(40, 40))
MouseMovedTo(pos=(50, 50))
MouseMovedTo(pos=(60, 60))
MouseMovedTo(pos=(70, 70))
MouseMovedTo(pos=(80, 80))
MouseMovedTo(pos=(90, 90))
MouseMovedTo(pos=(100, 100))
MouseMovedTo(pos=(110, 110))
MouseMovedTo(pos=(120, 120))
MouseMovedTo(pos=(130, 130))
MouseMovedTo(pos=(140, 140))
MouseMovedTo(pos=(150, 150))
MouseMovedTo(pos=(160, 160))
MouseMovedTo(pos=(170, 170))
MouseMovedTo(pos=(180, 180))
MouseMovedTo(pos=(190, 190))
MouseMovedTo(pos=(200, 200))
MouseMovedTo(pos=(210, 210))
MouseMovedTo(pos=(220, 220))
MouseMovedTo(pos=(230, 230))
MouseMovedTo(pos=(240, 240))
MouseMovedTo(pos=(250, 250))
MouseMovedTo(pos=(260, 260))
MouseMovedTo(pos=(270, 270))
MouseMovedTo(pos=(280, 280))
MouseMovedTo(pos=(290, 290))
MouseMovedTo(pos=(300, 300))
MouseMovedTo(pos=(310, 310))
MouseMovedTo(pos=(320, 320))
MouseMovedTo(pos=(330, 330))
MouseMovedTo(pos=(340, 340))
MouseMovedTo(pos=(350, 350))
MouseMovedTo(pos=(360, 360))
MouseMovedTo(pos=(370, 370))
MouseMovedTo(pos=(380, 380))
MouseMovedTo(pos=(390, 390))
MouseMovedTo(pos=(400, 400))
MouseMovedTo(pos=(410, 410))
MouseMovedTo(pos=(420, 420))
MouseMovedTo(pos=(430, 430))
MouseMovedTo(pos=(440, 440))
MouseMovedTo(pos=(450, 450))
MouseMovedTo(pos=(460, 460))
MouseMovedTo(pos=(470, 470))
MouseMovedTo(pos=(480, 480))
MouseMovedTo(pos=(490, 490))


FINAL STATE:
GameState
It is reds turn.
model.time=None
The mouse is at (490, 490).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O


FINAL SCREEN:
...............................................................................
...............................W..W...........W................................
..............................WWWWW.WWWW.WWW.W.WWW.............................
.................................RR............................................
...............................RRRRR...........................................
...............................RRRRRR..........................................
..............................RRRRRRR..........................................
..............................RRRRRRR..........................................
...............................RRRRR...........................................
.................BBBBBBBBBBBBBBBRRRBBBBBBBBBBBBBBBBBBBBBBBBBBBB................
.................BB....BB....BBB....BB....BB....BBB....BB....BB................
.................B.....B......B.....B......B.....B......B.....B................
...............................................................................
...............................................................................
.................B.....B......B.....B......B.....B......B.....B................
.................BB....BB....BBB...BBB....BBB...BBB....BB....BB................
.................BB...BBBB..BBBB...BBBB..BBBB...BBBB..BBBB...BB................
.................B.....B......B.....B......B.....B......B.....B................
.................B............................................B................
...............................................................................
.................B............B..................B............B................
.................B.....BB.....B.....BB....BB.....B.....BB.....B................
.................BBB.BBBBB..BBBBB.BBBBB..BBBBB.BBBBB..BBBBB.BBB................
.................B.....BB....BB.....BB....BB.....BB....BB.....B................
.................B............B.....B......B.....B............B................
...............................................................................
...............................................................................
.................B.....B......B.....B......B.....B......B.....B................
.................BB...BBB....BBB...BBB....BBB...BBB....BBB...BB................
.................BB...BBB....BBB...BBB....BBB...BBB....BBB...BB................
.................B.....B......B.....B......B.....B......B.....B................
...............................................................................
...............................................................................
.................B............B.....B......B.....B............B................
.................B.....BB....BB.....BB....BB.....BB....BB.....B................
.................BBB.BBBBB..BBBBB.BBBBB..BBBBB.BBBBB..BBBBB.BBB................
.................B.....BB.....B.....BB....BB.....B.....BB.....B................
.................B............B..................B............B................
...............................................................................
.................B............................................B................
.................B.....B......B.....B......B.....B......B.....B................
.................BB...BBBB..BBBB...BBBB..BBBB...BBBB..BBBB...BB................
.................BB....BB....BBB...BBB....BBB...BBB....BB....BB................
.................B.....B......B.....B......B.....B......B.....B................
...............................................................................
...............................................................................
.................B.....B......B.....B......B.....B......B.....B................
.................BB....BB....BBB....BB....BB....BBB....BB....BB................
.................BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB................
...............................................................................
...............................................................................
...............................................................................
...............................................................................
...............................................................................
...............................................................................
...............................................................................
...............................................................................


SIMULATION LOG:
[SIMULATION STARTING]
===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (550, 400).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(0, 0))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (0, 0).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(10, 10))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (10, 10).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(20, 20))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (20, 20).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(30, 30))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (30, 30).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(40, 40))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (40, 40).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(50, 50))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (50, 50).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(60, 60))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (60, 60).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(70, 70))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (70, 70).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(80, 80))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (80, 80).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(90, 90))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (90, 90).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(100, 100))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (100, 100).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(110, 110))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (110, 110).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(120, 120))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (120, 120).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(130, 130))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (130, 130).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(140, 140))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (140, 140).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(150, 150))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (150, 150).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(160, 160))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (160, 160).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(170, 170))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (170, 170).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(180, 180))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (180, 180).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(190, 190))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (190, 190).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(200, 200))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (200, 200).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(210, 210))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (210, 210).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(220, 220))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (220, 220).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(230, 230))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (230, 230).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(240, 240))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (240, 240).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(250, 250))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (250, 250).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(260, 260))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (260, 260).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(270, 270))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (270, 270).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(280, 280))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (280, 280).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(290, 290))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (290, 290).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(300, 300))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (300, 300).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(310, 310))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (310, 310).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(320, 320))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (320, 320).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(330, 330))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (330, 330).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(340, 340))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (340, 340).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(350, 350))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (350, 350).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(360, 360))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (360, 360).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(370, 370))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (370, 370).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(380, 380))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (380, 380).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(390, 390))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (390, 390).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(400, 400))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (400, 400).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(410, 410))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (410, 410).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(420, 420))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (420, 420).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(430, 430))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (430, 430).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(440, 440))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (440, 440).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(450, 450))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (450, 450).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(460, 460))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (460, 460).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(470, 470))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (470, 470).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(480, 480))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (480, 480).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATING MSG=MouseMovedTo(pos=(490, 490))]

===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (490, 490).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATION ENDED]
